File: apps/mobile-legacy-rn/src/lib/access.ts
import { ROLES, type AuthUser, type ServiceTrip } from '@servis/shared';

/**
 * İstemci tarafı görünürlük kuralları (UX içindir; asıl kontrol backend'de).
 * - Şoför yalnızca kendisine atanmış servisleri görür.
 * - Yolcu yalnızca yolcusu olduğu servisleri görür.
 * - Yönetici/operasyon yalnızca kendi tenant'ının servislerini görür.
 * - Süper admin platform genelini görür.
 */
export function canViewTrip(
  user: AuthUser,
  trip: ServiceTrip,
  passengerTripIds: Set<string> = new Set(),
): boolean {
  // Tenant izolasyonu: super_admin hariç farklı tenant görülemez.
  if (user.role !== ROLES.SUPER_ADMIN && user.tenant_id !== trip.tenant_id) {
    return false;
  }
  switch (user.role) {
    case ROLES.SUPER_ADMIN:
      return true;
    case ROLES.COMPANY_ADMIN:
    case ROLES.OPERATIONS_MANAGER:
      return true; // aynı tenant
    case ROLES.DRIVER:
      return trip.driver_id === user.id;
    case ROLES.PASSENGER:
      return passengerTripIds.has(trip.id);
    default:
      return false;
  }
}

/** Kullanıcının bir kayda tenant açısından erişebilir olup olmadığı. */
export function assertSameTenant(user: AuthUser, recordTenantId: string): boolean {
  if (user.role === ROLES.SUPER_ADMIN) return true;
  return user.tenant_id === recordTenantId;
}

/** Bir kullanıcı için görünür servisleri süzer. */
export function scopeTrips(
  user: AuthUser,
  trips: ServiceTrip[],
  passengerTripIds: Set<string> = new Set(),
): ServiceTrip[] {
  return trips.filter((t) => canViewTrip(user, t, passengerTripIds));
}

File: apps/mobile/lib/data/access.dart
import '../core/constants/roles.dart';
import 'models/auth_models.dart';
import 'models/service_trip.dart';

/// İstemci tarafı görünürlük kuralları (UX içindir; asıl kontrol backend'de).
///
/// - Şoför yalnızca kendisine atanmış servisleri görür.
/// - Yolcu yalnızca yolcusu olduğu servisleri görür.
/// - Yönetici/operasyon yalnızca kendi tenant'ının servislerini görür.
/// - Süper admin platform genelini görür.
bool canViewTrip(AuthUser user, ServiceTrip trip, {Set<String> passengerTripIds = const {}}) {
  // Tenant izolasyonu: super_admin hariç farklı tenant görülemez.
  if (user.role != Role.superAdmin && user.tenantId != trip.tenantId) return false;

  switch (user.role) {
    case Role.superAdmin:
      return true;
    case Role.companyAdmin:
    case Role.operationsManager:
      return true; // aynı tenant
    case Role.driver:
      return trip.driverId == user.id;
    case Role.passenger:
      return passengerTripIds.contains(trip.id);
  }
}

/// Kullanıcının bir kayda tenant açısından erişimi.
bool assertSameTenant(AuthUser user, String recordTenantId) =>
    user.role == Role.superAdmin || user.tenantId == recordTenantId;

/// Bir kullanıcı için görünür servisleri süzer.
List<ServiceTrip> scopeTrips(AuthUser user, List<ServiceTrip> trips,
        {Set<String> passengerTripIds = const {}}) =>
    trips.where((t) => canViewTrip(user, t, passengerTripIds: passengerTripIds)).toList();
